File: 프로그래머스/LV2/코딩테스트 연습/2020 KAKAO BLIND RECRUITMENT/괄호 변환/괄호 변환.py
def flip(u):
    ret = ""
    for c in u:
        if c == '(': ret += ')'
        else: ret += '('
    return ret


def correct_parenthesis(s):
    stack = []
    top = -1
    for c in s:
        if top != -1 and stack[top] == '(' and c == ')':
            stack.pop()
            top -= 1
        else:
            stack.append(c)
            top += 1
    
    return True if top == -1 else False
        

def go(w):

    if w == "": return ""

    num_of_open, num_of_close = 0, 0
    s, u, v = "", "", ""

    for i in range(len(w)):
        if w[i] == '(': num_of_open += 1
        elif w[i] == ')': num_of_close += 1
        
        if num_of_open == num_of_close:
            u = w[:i+1]
            v = w[i+1:]
            break

    if correct_parenthesis(u):
        u += go(v)
        return u
    else:
        s = "(" + go(v) + ")"
        s += flip(u[1:-1])
        return s

def solution(p):
    if correct_parenthesis(p):
        return p

    return go(p)

File: 프로그래머스/LV2/코딩테스트 연습/2020 KAKAO BLIND RECRUITMENT/괄호 변환/괄호 변환.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

string flip(string u) {
    string ret = "";
    for (char c : u) {
        if (c == '(')ret += ")";
        else ret += "(";
    }
    return ret;
}

bool correct_parenthesis(string s) {
    stack<char> stack;
    for (char c : s) {
        if (!stack.empty() && stack.top() == '(' && c == ')')stack.pop();
        else stack.push(c);
    }
    return stack.empty() ? true : false;
}

string go(string w) {
    if (w == "")return "";

    int num_of_open = 0, num_of_close = 0;
    string s = "", u = "", v = "";

    for (int i = 0; i < w.length(); i++) {
        if (w[i] == '(')num_of_open++;
        else num_of_close++;

        if (num_of_open == num_of_close) {
            u = w.substr(0, i + 1);
            v = w.substr(i + 1);
            break;
        }
    }

    if (correct_parenthesis(u)) {
        u += go(v);
        return u;
    }
    else {
        s += "(" + go(v) + ")";
        u = u.substr(1, u.length()-2);
        s += flip(u);
        return s;
    }
}

string solution(string p) {
    if (correct_parenthesis(p))
        return p;

    return go(p);
}